examples/common/server-cluster-shim/ServerClusterShim.cpp

```diff
@@ -0,0 +1,646 @@
+    {
+        return std::nullopt;
+    }
+
+    CHIP_ERROR err = aai->Read(path, encoder);
+
+    if (err != CHIP_NO_ERROR)
+    {
+        // Implementation of 8.4.3.2 (Incoming Read Request and Subscribe Request Action Processing)
+        // of the spec for path expansion
+        if (path.mExpanded && (err == CHIP_IM_GLOBAL_STATUS(UnsupportedRead)))
+        {
+            return CHIP_NO_ERROR;
+        }
+
+        return err;
+    }
+
+    // If the encoder tried to encode, then a value should have been written.
+    //   - if encode, assume DONE (i.e. FINAL CHIP_NO_ERROR)
+    //   - if no encode, say that processing must continue
+    return encoder.TriedEncode() ? std::make_optional(CHIP_NO_ERROR) : std::nullopt;
+}
+
+/// Attempts to write via an attribute access interface (AAI)
+///
+/// If it returns a CHIP_ERROR, then this is a FINAL result (i.e. either failure or success)
+///
+/// If it returns std::nullopt, then there is no AAI to handle the given path
+/// and processing should figure out the value otherwise (generally from other ember data)
+std::optional<CHIP_ERROR> TryWriteViaAccessInterface(const ConcreteDataAttributePath & path, AttributeAccessInterface * aai,
+                                                     AttributeValueDecoder & decoder)
+{
+    // Processing can happen only if an attribute access interface actually exists..
+    if (aai == nullptr)
+    {
+        return std::nullopt;
+    }
+
+    CHIP_ERROR err = aai->Write(path, decoder);
+
+    if (err != CHIP_NO_ERROR)
+    {
+        return std::make_optional(err);
+    }
+
+    // If the decoder tried to decode, then a value should have been read for processing.
+    //   - if decoding was done, assume DONE (i.e. final CHIP_NO_ERROR)
+    //   -  otherwise, if no decoding done, return that processing must continue via nullopt
+    return decoder.TriedDecode() ? std::make_optional(CHIP_NO_ERROR) : std::nullopt;
+}
+
+DataModel::AttributeEntry AttributeEntryFrom(const ConcreteClusterPath & clusterPath, const EmberAfAttributeMetadata & attribute)
+{
+    const ConcreteAttributePath attributePath(clusterPath.mEndpointId, clusterPath.mClusterId, attribute.attributeId);
+
+    using DataModel::AttributeQualityFlags;
+
+    DataModel::AttributeEntry entry(
+        attribute.attributeId,
+        BitFlags<DataModel::AttributeQualityFlags>{}
+            .Set(AttributeQualityFlags::kListAttribute, (attribute.attributeType == ZCL_ARRAY_ATTRIBUTE_TYPE))
+            .Set(DataModel::AttributeQualityFlags::kTimed, attribute.MustUseTimedWrite()),
+        RequiredPrivilege::ForReadAttribute(attributePath),
+        attribute.IsReadOnly() ? std::nullopt : std::make_optional(RequiredPrivilege::ForWriteAttribute(attributePath)));
+
+    // NOTE: we do NOT provide additional info for:
+    //    - IsExternal/IsSingleton/IsAutomaticallyPersisted is not used by IM handling
+    //    - IsSingleton spec defines it for CLUSTERS where as we have it for ATTRIBUTES
```

Comment:
That's because IsSingleton as used here is a Zigbee thing that should not exist in the SDK at all but got pulled in during the initial code landings and when people were trying to do a unified data model.  Just don't mention it here, please, and @andy31415 has a note to do a followup to remove it entirely.

See also https://github.com/project-chip/connectedhomeip/pull/39693#discussion_r2170150515